client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java

```diff
@@ -239,4 +243,31 @@ public void close() {
         connectionFactory.close();
     }
 
+    public void validateStartAndEndStreamCuts(final ReaderGroupConfig readerGroupConfig) {
+        readerGroupConfig.getEndingStreamCuts().entrySet().stream().filter(e -> !e.getValue().equals(StreamCut.UNBOUNDED))
+                .forEach(e -> {
+                    if (readerGroupConfig.getStartingStreamCuts().get(e.getKey()) != StreamCut.UNBOUNDED) {
+                        verifyStartAndEndStreamCuts(readerGroupConfig.getStartingStreamCuts().get(e.getKey()), e.getValue());
+                    }
+                });
+    }
+
+    private void verifyStartAndEndStreamCuts(final StreamCut startStreamCut, final StreamCut endStreamCut) {
+        final Map<Segment, Long> startPositions = startStreamCut.asImpl().getPositions();
+        final Map<Segment, Long> endPositions = endStreamCut.asImpl().getPositions();
+
+        //* verify both stream cut satisfy the "cover full key space" and "no key space overlap" in each of the streamcuts
+        Preconditions.checkArgument(controller.isStreamCutValid(startStreamCut).equals(true),
```

Comment:
these are remove calls being made to controller. you may want to chain these futures for optimization.
secondly, if a new api is introduced that takes two stream cuts, then the validation of individual cuts can happen as part of that single api call and you dont need to make two separate calls into controller.

Reply:
also, `controller.isStreamCutValid(endStreamCut)` returns a future.. you cannot compare it with a boolean

Reply:
Correct. There is no need to use `equals(boolean)`; just use ` == true` and the compiler would have caught this problem.

Also, if it's a Future, then this must be making a call over the wire. This will make this method very slow.

Reply:
Please write a test to cover this case.